pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/ConsistentDataPushUtils.java

```diff
@@ -0,0 +1,296 @@
+    String rawTableName = spec.getTableSpec().getTableName();
+    if (!uriToLineageEntryIdMap.isEmpty()) {
+      LOGGER.info("End consistent push for table: " + rawTableName);
+      endReplaceSegments(spec, uriToLineageEntryIdMap);
+    }
+  }
+
+  public static List<String> getMetadataSegmentsTo(Map<String, String> segmentUriToTarPathMap) {
+    List<String> segmentsTo = new ArrayList<>();
+    for (String segmentUriPath : segmentUriToTarPathMap.keySet()) {
+      String tarFilePath = segmentUriToTarPathMap.get(segmentUriPath);
+      String fileName = new File(tarFilePath).getName();
+      Preconditions.checkArgument(fileName.endsWith(Constants.TAR_GZ_FILE_EXT));
+      String segmentName = fileName.substring(0, fileName.length() - Constants.TAR_GZ_FILE_EXT.length());
+      segmentsTo.add(segmentName);
+    }
+    return segmentsTo;
+  }
+
+  public static Map<URI, URI> getStartReplaceSegmentUris(SegmentGenerationJobSpec spec, String rawTableName) {
+    Map<URI, URI> baseUriToStartReplaceSegmentUriMap = new HashMap<>();
+    for (PinotClusterSpec pinotClusterSpec : spec.getPinotClusterSpecs()) {
+      URI controllerURI;
+      try {
+        controllerURI = new URI(pinotClusterSpec.getControllerURI());
+        baseUriToStartReplaceSegmentUriMap.put(controllerURI,
+            FileUploadDownloadClient.getStartReplaceSegmentsURI(controllerURI, rawTableName,
+                TableType.OFFLINE.toString(), true));
+      } catch (URISyntaxException e) {
+        throw new RuntimeException("Got invalid controller uri - '" + pinotClusterSpec.getControllerURI() + "'");
+      }
+    }
+    return baseUriToStartReplaceSegmentUriMap;
+  }
+
+  public static Map<URI, String> startReplaceSegments(SegmentGenerationJobSpec spec,
+      Map<URI, List<String>> uriToSegmentsFrom, List<String> segmentsTo)
+      throws Exception {
+    Map<URI, String> uriToLineageEntryIdMap = new HashMap<>();
+    String rawTableName = spec.getTableSpec().getTableName();
+    Map<URI, URI> segmentsUris = getStartReplaceSegmentUris(spec, rawTableName);
+    AuthProvider authProvider = AuthProviderUtils.makeAuthProvider(spec.getAuthToken());
+    LOGGER.info("Start replace segment URIs: " + segmentsUris);
+
+    int attempts = 1;
+    long retryWaitMs = 1000L;
+
+    for (Map.Entry<URI, URI> entry : segmentsUris.entrySet()) {
+      URI controllerUri = entry.getKey();
+      URI startSegmentUri = entry.getValue();
+      List<String> segmentsFrom = uriToSegmentsFrom.get(controllerUri);
+
+      if (!Collections.disjoint(segmentsFrom, segmentsTo)) {
+        // This scenario could happen only when "disable.unique.segment.name" is enabled,
+        // i.e. newly generated segment name is the same as the existing one in the table. If so, we should
+        //    1) either clean up all the consistent push custom config from table config,
+        //    2) or remove "disable.unique.segment.name" to continue enabling consistent push.
+        String errorMsg =
```

Comment:
We should add the segment names to the log message here. Also since we are throwing exactly with the same message, should we log at all? same message will be in the trace even if we don't log  right?